### PR TITLE
rework(balance): Penguin rework

### PR DESCRIPTION
**Pull request by ziproot**
**Balance**

## Summary
In #9693, @Zitchas expressed interest in giving the integrated engines an energy and heat cost. To prevent merge conflicts I have not changed the Merganser, but I have made the Penguin cost energy and heat now.
EDIT: It now also produces its own energy and creates heat from that process at the same efficiency rate as a millenium cell. You can pretend it's using materials just like you pretend nuclear reactors use materials. I made it produce enough energy to power its thrusters. All tests below should be about the same as they were before, as they were tested with lower energy generation. It still isn't possible run donuts indefinitely, however, it takes 30 seconds to deplete batteries and two seconds to recharge them, and as donuting is more of a "just for fun" activity I think having to wait two seconds to do it again for another 30 is not an issue.
EDIT: This PR now also buffs thrust, reverse thrust, and turn by 1.5× as this was not done in the mass rebalance.
EDIT: This PR now also buffs heat dissipation to compensate for the worsened ability to withstand heat in the ember waste.

## Screenshots
Before
before

After
after

## Testing Done

Tested to make sure the following still worked:
- The Penguin can recover from an ion storm quickly (OK)
- The Penguin can circle around an enemy without losing power (OK)
- The Penguin can gather around your flagship without losing power (OK)
- The Penguin can tail a ship for the purposes of scanning without losing power (OK)
- The Penguin can dodge attacks from the archon without losing power (OK)
- The Penguin can navigate hyperspace with the cloaking device active and not lose power (OK)
- The Penguin can land with the cloaking device active (OK)*
- The Penguin can flee while recharging shields and not lose power (OK)

Tests continue to pass as of https://github.com/endless-sky/endless-sky/pull/9704/commits/b330c613e7d24eed8a0f46865891bdf616be75d9

*This was not the original test, which also required the Penguin to land without losing power, however, even with the inefficient landing corrected and the mass rework buffing thrust/energy, the Penguin still cannot land cloaked without losing power, even on master. In other words, the test was updated so that it still accurately tests for regressions.

## Save File
This save file can be used to test these changes: 
[Remnant Test~Penguin.txt](https://github.com/endless-sky/endless-sky/files/15487787/Remnant.Test.Penguin.txt)

### CODE REVIEW

**Comment by ziproot** [comment]
The run failed but that has nothing to do with this PR:
```
Warning: Failed to download action 'https://api.github.com/repos/lukka/get-cmake/tarball/2654d8ee382b9b6cbbfe6487653b8629b4e062c8'. Error: nodename nor servname provided, or not known (codeload.github.com:443)
Warning: Back off 15.484 seconds before retry.
Warning: Failed to download action 'https://api.github.com/repos/lukka/get-cmake/tarball/2654d8ee382b9b6cbbfe6487653b8629b4e062c8'. Error: nodename nor servname provided, or not known (codeload.github.com:443)
Warning: Back off 22.893 seconds before retry.
```

File: data/remnant/remnant ships.txt [hunk 1]
```diff
@@ -1108,8 +1108,14 @@ ship "Penguin"
 		"ion resistance" 0.05
 		"scramble resistance" 0.05
 		"thrust" 4.2
+		"thrusting energy" 0.65
+		"thrusting heat" 0.6
```
Comment (Zitchas) [inline data/remnant/remnant ships.txt:1112]:
```suggestion
		"thrusting energy" 0.31
		"thrusting heat" 0.29
```

As a built-in system, it dispenses with a lot of inefficiency that is required to support customization, modularity, etc. As such, the least efficient existing Remnant system (the Anvil) should be the maximum energy/heat cost per unit. There's no reason why such a system should be twice as inefficient as the anvil is.

File: data/remnant/remnant ships.txt [hunk 2]
```diff
@@ -1108,8 +1108,14 @@ ship "Penguin"
...
 		"turn" 80.0
+		"turning energy" 0.25
+		"turning heat" 0.25
```
Comment (Zitchas) [inline data/remnant/remnant ships.txt:1115]:
```suggestion
		"turning energy" 0.15
		"turning heat" 0.16
```

This brings it in line with the Anvil.

File: data/remnant/remnant ships.txt (hunk 1 above)
Comment (ziproot) [reply data/remnant/remnant ships.txt]: I got around 0.62 thrusting energy and 0.59 thrusting heat if using the anvil class engine efficiencies, so I've lowered the costs to 0.6 energy and 0.55 heat.

File: data/remnant/remnant ships.txt (hunk 2 above)
Comment (ziproot) [reply data/remnant/remnant ships.txt]: Done

**Comment by Zitchas** [comment]
Tentatively looks good, haven't tested it yet, though.

**Review by OcelotWalrus — APPROVED** [review]

**Review by Quantumshark — APPROVED** [review]

**Comment by warp-core** [comment]
I'm not sure how the 0.15 energy and heat costs for turning have been reached. The comments (https://github.com/endless-sky/endless-sky/pull/9704#pullrequestreview-1836316132 and https://github.com/endless-sky/endless-sky/pull/9704#discussion_r1461767242) reference the Anvil, but the current values here result in significantly higher efficiency than the Anvil class engines. It's actually more efficient than all the Remnant engines. If the goal is to hit similar efficiency to the Anvil, I would expect about 0.22 turning energy and about 0.24 turning heat.

**Comment by ziproot** [comment]
The issue was I plugged in 60 for the turn instead of 80. I used the merganser's turn instead of the penguin. Oops

**Comment by Zitchas** [comment]
It isn't supposed to have the same as the anvil, rather the anvil was the baseline. As in that is the minimum possible efficiency. It should most definitely be better than that, since it has all the before mentioned opportunities for efficiency.

**Comment by ziproot** [comment]
It *is* better than that. It has 0.20 turning heat/energy, not 0.24/0.22 turning heat/energy. I just messed up the math.

**Comment by Zitchas** [comment]
Yea, I was just explaining to Warp-core.

**Comment by ziproot** [comment]
No worries.

**Comment by warp-core** [comment]
Here's what I get for the energy and heat efficiency of the Remnant engines:

property | Anvil | Crucible | Forge | Smelter
-- | -- | -- | -- | --
thrust/energy | ~6.73 | ~7.88 | ~7.73 | ~7.48
thrust/heat | ~7.08 | 8.275 | 8 | ~7.41
reverse/energy | ~6.87 | 7.75 | ~7.73 | ~7.51
reverse/heat | ~7.10 | ~8.27 | ~7.97 | ~7.40
turn/energy | 352 | ~345 | ~377 | ~407
turn/heat | 336 | ~326 | ~332 | ~342

With 0.2 energy and heat and 80 turn, the Penguin has 400 thrust/energy and thrust/heat. (The previous values of 0.15 put it over 500.)
This puts the Penguin only slightly behind the Smelter in terms of energy efficiency (still quite far ahead the Anvil), and well ahead of all four of the others in terms of heat efficiency.

**Comment by ziproot** [comment]
I didn't realize that the anvil was actually pretty good for turning efficiency in terms of energy/heat. Using your suggestions puts it ahead of the actual worst engines in turning (crucible).

**Comment by alexrovw** [comment]
Genuinely don't mean to sound dismissive here, but what is the need for this change? Is the Penguin too powerful? Too good at something? Why are we making it worse?

I playtested the proposed changes. With the default outfits, the Penguin now burns through its meager energy storage pretty quickly just by flying around.

New Penguin:

https://github.com/endless-sky/endless-sky/assets/98854112/97ed8d99-0e93-4159-a672-cbbef16a50d3

Old Penguin:

https://github.com/endless-sky/endless-sky/assets/98854112/5aad8438-1d63-4fec-8646-b482d8b0de0e

You could add a crystal capacitor, but then you've eaten up almost all of your free outfit space for dubious benefit. As the Penguin is intended to be a straight upgrade to the Puffin, this kind of thing shouldn't happen. The ship should be a visible improvement over the outdated model but also stand on its own as a usable ship.

With or without the change, you can see that having the cloak on results in very quick drain of the battery. Thus, the presence of a built-in cloak doesn't necessitate this change either.

The Penguin isn't exactly an overpowered ship. For what reason are we re-balancing the built-in engines? I'm trying to assume good faith here, but from the outside looking in, it _seems_ like we're just making X numbers closer to Y numbers without any actual gameplay improvement following suit.

**Comment by ziproot** [comment]
> Genuinely don't mean to sound dismissive here, but what is the need for this change? Is the Penguin too powerful? Too good at something? Why are we making it worse?
If you take out the engines, it has infinite thrust/heat, infinite thrust/energy, infinite thrust/space, infinite reverse thrust/heat, infinite reverse thrust/energy, infinite reverse thrust/space, infinite turn/heat, infinite turn/energy, and infinite turn/space. That's too powerful if you ask me.

**Comment by warp-core** [comment]
> If you take out the engines, it has infinite thrust/heat, infinite thrust/energy, infinite thrust/space, infinite reverse thrust/heat, infinite reverse thrust/energy, infinite reverse thrust/space, infinite turn/heat, infinite turn/energy, and infinite turn/space.

While this is a true statement, I'm not sure it's actually a very useful way of looking at it.
That thrust and turning may not have an obvious cost, but I think it's worth considering how much outfit space it is worth. That is to say, how much outfit space would it take to add these attributes to the ship with some relevant, normal outfits. I'm going to use the Crucible-class engines the ship has installed as the reference.
The thrust on the Penguin is about 1/4 of what the Crucible provides, the reverse about 1/3. Call it a quarter of a Crucible thruster. 5 tonnes.
The turn is about 2/11 of a Crucible, so about 2.5, I'll round up to 3 tonnes.
A quarter of the Crucible's thrust energy and heat costs is about 30.
A sixth of the steering costs is about 14.
I'll just say about 45 energy and heat. How much outfit space would it take to generate 45 energy and dissipate 45 heat (plus the heat from the energy generation)?
It's a little tricky because the relevant cooling outfit also consumes energy. I think about 2/7 of a Millenium Cell, and 1/2 of a Thermoelectric Cooler, giving about 60 energy generation and about 100 heat generation from the Cell, for about 8 tonnes, and 240 cooling and 12 energy use from the cooler for 2 tonnes.

So, the engines themselves would cost 8 tonnes to match and the support outfits would be about 10 tonnes.

In general, I consider the outfit space to be worth somewhat more than the attributes you could get from it with any particular set of outfits. Having the outfit space offers more flexibility. If you have outfit space, you can put whatever you want in there, but with the attribute on the ship itself, you have no choice.
Also, there just not being any costs to thrusting or turning is additionally worse than having those costs, having some more outfit space, and using it to add more energy generation and cooling, because having those outfits would provide additional energy generation and cooling that can be used by other things while not thrusting and turning.
However, it would be rather difficult to add such small amounts of energy generation and cooling, Outfits this small just don't exist. You'd have to substitute some installed outfits for larger ones, likely ending up with considerably more energy generation and cooling than you actually need. This does increase the value of having the attributes on the ship.

Overall, I don't really think the lack of energy usage or heat production from the built in engines is a problem.

**Comment by ziproot** [comment]
I guess I can close it then, but I don't really follow what you are saying

**Comment by ziproot** [comment]
I'm reopening it because I'm still not sure what @warp-core wants.

**Comment by ziproot** [comment]
It should now take 30 seconds to run out of energy if using full thrusters and turn at once. Additionally, this adds 2 integrated outfit space in the form of the supercapacitor making it 8 (engines) + 2 (supercapacitor) - 10 (powering and cooling engines) = 0

**Comment by Zitchas** [comment]
You should also be adding energy generation, too. It's a living being, after all. Like all living things, it produces it's own energy for moving around. This was originally implicit in how it was laid out (basically assuming for simplicity that energy used = energy produced, and thus leaving it all out to avoid complicating outfitting), but if you want to add energy expenditures, then adding the whole setup is then necessary.

**Comment by ziproot** [comment]
> Like all living things, it produces it's own energy for moving around

I don't produce my own energy...
Do you mean photosynthesis (built in solar collection)?

**Comment by Zitchas** [comment]
Yes, you do. You process some raw materials (food, water, oxygen, etc) and generate energy with them. 

Ideally, if this is actually all being mechanically implemented, then it should be converting a low amount of fuel into energy. Or if one wishes to be more sporadic and different having ion dissipation generate energy (as in, as it dissipates ionization, it produces energy from it). At tiny rates, of course.

**Comment by alexrovw** [comment]
> > Like all living things, it produces it's own energy for moving around
> 
> I don't produce my own energy... Do you mean photosynthesis (built in solar collection)?

You do. All living things produce [biochemical energy](https://www.sciencedirect.com/topics/engineering/biochemical-energy#:~:text=Biochemical%20energy%20is%20manipulated%20by,microorganisms%20to%20produce%20those%20fuels.). Photosynthesis is the mechanism by which plants do so. Most animals, unless they're parasites, typically do it by consuming and processing food.

Seems like if we're going to roll with the living ship mechanics, they could have a little bit of energy generation like Zitchas said.

**Comment by ziproot** [comment]
When I read "produces it's own energy," I read it as doing so self-sufficiently (without needing to ingest chemicals to break down). I was thinking of the Endless Sky nuclear reactors which seemingly make energy out of thin air. However, I forgot that in the real world nuclear reactors also use raw materials. Endless Sky in that case hand waves the material cost to keep complexity down, so I don't think a material requirement is necessary for the penguin's built-in energy generation.

(Also, plants use cellular respiration like all other living things on Earth to make energy, photosynthesis is how they make glucose to store for cellular respiration later)

**Review by Quantumshark — APPROVED** [review]
The extra energy generation seems reasonable. The Penguin is a unique mission reward ship which currently doesn't have a lot going for it, especially compared to the other, better gaslining ships which unlock later on in the Remnant campaign.
So, I think it's a good thing for it to get a little bit of bonus energy generation. It's only the equivalent of about three space anyways, so it's not like it's a particularly major buff.

**Comment by ziproot** [comment]
Update: slightly nerfed builtin energy generation so that outfit space needed to account for energy/heat from the builtin engines is equal to the outfit space the free engines would normally take up

**Comment by alexrovw** [comment]
> Update: slightly nerfed builtin energy generation so that outfit space needed to account for energy/heat from the builtin engines is equal to the outfit space the free engines would normally take up EDIT: Basically, the energy generation is the equivalent of two outfit space rather than three

I'm not sure I understand the reason for this extra nerf. Even with the pre-nerf energy generation you provided (`0.6`) the Penguin runs out of energy doing donuts, which is a regression from the previous Penguin UX before this PR. With it lowered to `0.4`, it runs out of energy even faster. What's the benefit of doing this?

**Comment by ziproot** [comment]
> I'm not sure I understand the reason for this extra nerf. Even with the pre-nerf energy generation you provided (0.6) the Penguin runs out of energy doing donuts, which is a regression from the previous Penguin UX before this PR. With it lowered to 0.4, it runs out of energy even faster. What's the benefit of doing this?

I don't see how that could be the case, as from my tests it never lost power when thrusting alone, and turning uses less than 0.4 energy per frame. I could re-test though.

**Comment by ziproot** [comment]
I have re-tested it and it takes 20 seconds to fully deplete batteries when using both thrusters and steering.

I am fine with this, as this is more of a rework now and less of a nerf:
-When just thrusting, just turning, idling, or otherwise using it in most common cases, energy generation is buffed
-When thrusting and turning, energy generation is nerfed

The extra nerf I did was based on warp core's analysis. I don't fully understand it, but it seems like warp core wants the "free" engine space to be balanced with the outfit space needed to accomodate the built-in engines. I am willing to revert that to what it was before now that I know more about what's going on.

**Comment by ziproot** [comment]
@warp-core how does this feel?

**Comment by ziproot** [comment]
Updating because of #8156 being fixed in master.

**Comment by ziproot** [comment]
Tested and the penguin can now land with the cloak active.

**Comment by ziproot** [comment]
Merging to make sure spellcheck doesn’t bug

**Comment by Zitchas** [comment]
Did some flying around in the Ember Waste, and, well, it works. The heat levels are a fair bit higher than I like, so I'd appreciate them being decreased somewhat. Or some built in cooling added to complete the ecosystem you've added here.

**Comment by ziproot** [comment]
I can nerf energy + heat generation, but I don’t see why I should just buff cooling for a non-combat ship unless the ship is overheating.

The ship already “naturally” cools in the form of heat dissipation. I could increase heat dissipation slightly (to 0.95) but I would require @Quantumshark’s approval.

**Review by Zitchas — DISMISSED** [review]
The ship was balanced according to its stats previously. You can't just pick and choose, you either add the whole ecosystem, or you add nothing. It operates in a high heat environment, with enemies that use heat generating weapons. "It still doesn't overheat while moving" is *not* a "close enough" metric.

I'm not particularly picky how you solve it. There's a lot of options. Raising the heat dissipation (it is on the low side for a ship of its mass & size), giving it a very small amount of passive cooling, reducing the heat generating elements, a small amount of everything... Lots of options. 

As for your comment about not exceeding 0.95:

We have eight constructed ships in the game that have heat dissipation >= 1, including one as high as 27, and three biological ships with heat dissipation > 1, including two as high as 14. So don't feel restricted by the value "1". Some of those ships, like the model 8, are several times larger than the penguin. Although that being said, I don't think it should be necessary to follow the examples of those double digit numbers, or even significantly exceed 1, should you choose to go that route.

**Comment by Zitchas** [comment]
As a secondary thought: It's precisely the none-combat ships that *should* be considered for thermal buffs. Heat is an important limitation as part of the combat balance. For ships that aren't predominantly combat ships, it's not necessary to use it to reign them in.

**Comment by ziproot** [comment]
Marking as draft so I can address the changes caused by the mass rework.

**Comment by ziproot** [comment]
The mass rework changes caused the ratios in Remnant engines to change, due to thrust being buffed and energy/heat being unchanged, however, integrated thrust was unchanged, so I have lowered the energy and heat costs to match the new ratios, and lowered the energy and heat generation to match the lowered energy costs. This should result in a ship that overall produces less heat, addressing concerns raised by @Zitchas.

I would like approvals from @Quantumshark to test balance and @alexrovw to test if the ship still fares well after extended use.

**Comment by Quantumshark** [comment]
The Penguin's built-in engines should have received the same +50% to thrust and turn (and reverse thrust) in #9616 that all other engines in the game received. I don't know why they didn't, I must have missed it. You could fix that here, or I could fix it in a separate PR.

As for the issue about heat, the problem here I think is that the ember waste heat uses relative heat damage, and that scales off a ship's heat capacity, not off the total heat it can handle. The higher mass and lower heat capacity ships received in #9616 will have made all ships that were changed more vulnerable to relative heat damage. To fix this, all instances of relative heat damage should be divided by 1.5, as 1.5 times the prior mass and a division of the heat dissipation by 1.5 are the new 'baseline'. I'll open a PR to do that.

**Comment by ziproot** [comment]
This PR is dealing with energy and heat, so I think a separate PR that buffs thrust and reduces relative heat damage would be the best approach. If this PR is merged first, then heat and energy usage can be changed in that PR, otherwise, when I update again I can revert the changes to heat and energy usage.

**Comment by ziproot** [comment]
I think the issue of too much heat in the ember waste has been solved, but I will need to update the save file again.

**Comment by Zitchas** [comment]
I'll remind you that there is no objective balance. Your statements assume there was an imbalance as a factual statement, whereas it is actually just your subjective opinion.

The PR is, in my opinion, about adding detail. When I created the ship I looked at it and decided to leave off it off because a) they are small values, and b) adding small heat production and energy consumption and heat dissipation and energy production to offset those was  just taking an equation A = B and complexifying it to A-2*C-5 = B-2*C-5. No change. 

As such, I support this PR in the matter of adding detail. Detail is good. While I didn't feel the *need* to add that complexity to start with, I have no objection to adding it if someone wants to put in the work of doing the full circle.

What I dislike is the assumption that you have revealed that underlying this is the routine and regular intent that seems to crop up every time something touches existing Remnant ship stats: "How can this be used to make this ship worse?"

And, to be blunt, I see no justification for this. The only grounds I've seen mentioned so far is people making random guesses about how much hull mass it would "normally have" and try to work back how much mass accounts for various capabilities. Which is making so many guesses and assumptions that it really shouldn't be used for anything other than speculative wiki conspiracy theories.

**Comment by Quantumshark** [comment]
I don't have any issue with upping the Penguin's heat dissipation slightly. Unlike some other Remnant ships, the discussion of which you've alluded to, the Penguin's performance isn't excessive. I appreciate and agree with Azure's stance on those other ships, but nerfing the Penguin on their behalf isn't really going to make things more balanced.

In terms of the actual value change, increasing the heat dissipation from 0.82 to 0.919 would increase the maximum tolerable heat by as much as the total heat generated while thrusting and turning simultaneously has increased (including the increased idle heat), which would make the Penguin always as well off or better relative to where it is without the changes in this PR. So, there's no need to go past 0.92 with the heat dissipation, for certain. I'd suggest upping it to 0.88, since it's not going to be thrusting and turning simultaneously, and that's enough of a maximum heat increase to offset the increased heat from the idle generation and thrusters _or_ turning, leaving it slightly better off when idling and slightly worse off when thrusting and turning simultaneously.

**Comment by Zitchas** [comment]
Just to be clear, I was assuming that about you, not ziproot. I have zero assumptions about their intentions with this being anything other than, as stated, fleshing out the Penguin with the entire heat and energy cycle. 

The whole point of the built-in values was to establish that - no matter what the player does - the ship can still move and do stuff on its own. I hope no-one leaves the penguin without real thrusters, but just in case, this will enable future story involving it to always make sense, even when players are cruel. 

You keep talking about "Zero cost" and I find it both amusing and annoying. You don't know what tradeoffs I made it creating it. The assumption *appears* to be that I make none. 

Why yes, something purpose built into a specific ship, hardwired in, doesn't exactly match up with the performance of outfits that have been in production for decades and have to be built to be compatible with every ship configuration system in the galaxy, and entirely self contained. Purpose built and hardwired in do confer a number of advantages, and then there's that other thing: Innovation. Tech changes from year to year, never mind century to century. Things don't stay the same.

**Comment by ziproot** [comment]
I have updated the heat dissipation, so the Penguin can now “naturally” produce energy and cool itself down while thrusting or turning, as per @Quantumshark’s recommendation. After running tests, the Penguin seems to be about at half of its capacity when spinning donuts, which is about where it was before. Also keep in mind that this is a corner case where both thrust and turn are at their most extreme values (100% each), and most “normal” play (e.g. every test case) should result in *improved* cooling.

**Comment by Zitchas** [comment]
I can't test it right now, but at a theoretical level it looks ok.

**Comment by ziproot** [comment]
Just to clarify, I made this PR because Zitchas agreed with a comment by Hurleveur saying the Merganser should not be able to thrust/turn without any energy or heat usage, so for consistency, I did the same to the Penguin, after making sure Azure was OK with it. I then added energy generation at alexrovw’s request, as I did not realize how low the default energy storage was, and have repeatedly made sure that the changes I made would not break the balance of the game, which is why I have Quantumshark reviewing my changes. 

I have recently buffed heat dissipation because the mass rework meant that the Penguin now had to retroactively work harder in order to produce the same thrust, and the heat dissipation was adjusted due to the mass changes, however, these engines meant that the Penguin would now have had better heat dissipation had it been made with the increased work requirement in mind. Also, I again made sure this would not suddenly make the Penguin much more powerful than it was before, which is why I again made sure Quantumshark was fine with the change first.

However, I am doing this to address oversights. I am not doing this to make the Penguin exactly the same as it currently is. **The PR was made to address the Penguin getting to move for free, something Zitchas appeared to think should not be the case.** All other changes were adjustments to that fact. If it still basically can move for free, then this PR does nothing and it is just a waste of everyone’s time. However, **the Penguin still does not move for free with this PR, even with the changes to energy generation and heat dissipation**, as shown by it being buffed in some cases and nerfed in others, and I would like it to remain that way.

Thank you.

**Review by Quantumshark — APPROVED** [review]

**Comment by TheGiraffe3** [comment]
@Quantumshark will you remove the `to be merged` label?
